validator/src/journal/block_manager.rs

```diff
@@ -398,13 +402,158 @@ impl BlockManagerState {
 #[derive(Default, Clone)]
 pub struct BlockManager {
     state: Arc<BlockManagerState>,
+    persist_lock: Arc<RwLock<()>>,
 }
 
 impl BlockManager {
     pub fn new() -> Self {
         BlockManager::default()
     }
 
+    fn transaction_index_contains(
+        &self,
+        index: &Box<IndexedBlockStore>,
+        block_id: &str,
+        id: &str,
+    ) -> Result<bool, BlockManagerError> {
+        let index = index.as_ref();
+        if TransactionIndex::contains(index, id)? {
+            let block = TransactionIndex::get_block_by_id(index, id)?
+                .expect("BlockStore updated during transaction index check");
+            let head = index
+                .get(&[block_id])?
+                .next()
+                .expect("BlockStore updated during transaction index check");
+            if block.block_num <= head.block_num {
+                Ok(true)
+            } else {
+                Ok(false)
+            }
+        } else {
+            Ok(false)
+        }
+    }
+
+    fn batch_index_contains(
+        &self,
+        index: &Box<IndexedBlockStore>,
+        block_id: &str,
+        id: &str,
+    ) -> Result<bool, BlockManagerError> {
+        let index = index.as_ref();
+        if BatchIndex::contains(index, id)? {
+            let block = BatchIndex::get_block_by_id(index, id)?
+                .expect("BlockStore updated during batch index check");
+            let head = index
+                .get(&[block_id])?
+                .next()
+                .expect("BlockStore updated during batch index check");
+            if block.block_num <= head.block_num {
+                Ok(true)
+            } else {
+                Ok(false)
+            }
+        } else {
+            Ok(false)
+        }
+    }
+
+    pub fn contains_transaction(
+        &self,
+        block_id: &str,
+        id: &str,
+    ) -> Result<bool, BlockManagerError> {
+        let _lock = self
+            .persist_lock
+            .read()
+            .expect("The persist RwLock is poisoned");
```

Comment:
Can we just acquire a read lock on block_by_block_id?

Reply:
There is a deadlock that would happen. During splitting up the locks task, I made sure to have an precedence of locks, so that if an overlap of scope happened as happens in `put` and `unref`, `references_by_block_id` is locked first and then `block_by_block_id`, so there is always the same ordering of locks. In the `contains_*` functions, there is a call to `.branch` which calls `ref` which locks the references_by_block_id`, so we would have a `block_by_block_id` then `references_by_block_id` locking order, and a recipe for a deadlock.

Reply:
I think using a separate empty lock is actually more confusing than acquiring the lock around block_by_block_id. Acquiring the lock around block_by_block_id is a hard guarantee that nothing is going to be modifying which blocks are stored, whereas acquiring this empty lock is a soft guarantee that nothing is going to be modifying which blocks are being persisted, provided we always acquire the persist lock before persisting. Using an empty lock side steps the way Rust locks are designed to work. We are able to get away with it here, because we have the trait bound on BlockStore that it is thread-safe already. It might be cleaner to wrap BlockStore in an RwLock and remove the Sync trait bound, which I think would be the equivalent of this that doesn't require using an empty lock.

Reply:
After spending some time thinking about it, I am okay with using persist_lock here. I think there are probably better solutions, but I can't think of what they are and this should work.